### PR TITLE
Towards s390x support

### PR DESCRIPTION
**Pull request by rposts**
This PR partially enables s390x architecture support for Swift.

Specifically:

1. Modifies existing s390x SPARE_BIT_MASK
2. Adds s390x SWIFT_ABI_S390X_OBJC_RESERVED_BITS_MASK and supporting logic
3. Adds CLongDouble support on s390x
4. Updates isValidPointerForNativeRetain() function for s390x

### CODE REVIEW

**Review by millenomi — COMMENTED** [review]
I’m a little confused; Apple doesn’t ship an ObjC runtime on s390x. Can you comment on what’s up a little?

**Comment by rposts** [comment]
@millenomi - these changes are to address several validation test-case failures and continuation of earlier commits -e.g.: https://github.com/apple/swift/commit/85fde8b1fb363c97aec1b7b2ff32ce7a2f0da4ad

OBJC related changes are mostly addressing test case failures (>500 in number otherwise).

**Comment by jckarter** [comment]
What are those test failures? Ideally the ObjC defines would not impact platforms without an ObjC runtime. `BridgeObject`'s masking behavior may still be entangled with these masks, though. cc @milseman

**Comment by milseman** [comment]
I'm not familiar with the current status quo, but @mikeash has been refactoring some of this logic. BridgeObjects still need to work on Linux, so I suspect that the ObjC defines shouldn't be necessary, just the Swift ones.

**Comment by rposts** [comment]
@jckarter - here is a snippet of failures after restoring SWIFT_ABI_DEFAULT_OBJC_RESERVED_BITS_MASK :

```
 #define _swift_abi_ObjCReservedBitsMask                                        \
-  (__swift_uintptr_t) SWIFT_ABI_S390X_OBJC_RESERVED_BITS_MASK
+  (__swift_uintptr_t) SWIFT_ABI_DEFAULT_OBJC_RESERVED_BITS_MASK
...
    Swift(linux-s390x) :: stdlib/mmap.swift
    Swift(linux-s390x) :: stdlib/ReverseCompatibility.swift
    Swift(linux-s390x) :: stdlib/Strideable.swift
    Swift(linux-s390x) :: stdlib/Reflection.swift
    Swift(linux-s390x) :: stdlib/StringAPI.swift
    Swift(linux-s390x) :: stdlib/UnsafeRawBufferPointer.swift
...
    Swift(linux-s390x) :: Evolution/test_superclass_methods.swift
    Swift(linux-s390x) :: StdlibUnittest/CrashingTests.swift
    Swift(linux-s390x) :: Evolution/test_class_add_virtual_method.swift
    Swift(linux-s390x) :: StdlibUnittest/Stdin.swift
    Swift(linux-s390x) :: StdlibUnittest/CommandLine.swift
....
    Swift(linux-s390x) :: IRGen/objc_simd.sil
    Swift(linux-s390x) :: IRGen/errors.sil
    Swift(linux-s390x) :: Index/Store/output-failure.swift

  Expected Passes    : 9135
  Expected Failures  : 77
  Unsupported Tests  : 1386
  Unexpected Failures: 546

```
So it appears that test cases are heavily dependent on how these masks are defined

**Comment by mikeash** [comment]
#19540 shuffled a bunch of this stuff around about a week ago. The ObjC masks shouldn't be needed if ObjC interop isn't available, but in practice it looks like they are. In particular, we unconditionally incorporate the ObjC reserved bits into the `unTaggedNonNativeBridgeObjectBits` constant:

https://github.com/apple/swift/blob/master/stdlib/public/runtime/SwiftObject.mm#L495

Ideally, we shouldn't have the ObjC masks around at all if interop isn't available. I think the best solution would be to put the `heap_object_abi` ObjC constants at the bottom of `HeapObject.h` inside `#if SWIFT_OBJC_INTEROP`, and then cleaning up any place where those constants are used outside such a section. I'm not sure if that's beyond the scope of this work, though.

**Comment by rposts** [comment]
I think this area has seen quite a bit of refactoring in recent years. https://github.com/apple/swift/commit/3d04fb5eacbf197b04a776dbc71a7e1e2b05b07a#diff-ac47cb08ca19bd2108eaf8cb32ddcbc4 appears to have moved bit masks out of Metadata.h.

Scope of this PR is more on s390x big-endian enablement.  I wonder if broader cleanup can be attempted elsewhere to keep the scope separate with correct SME involved.

**Comment by millenomi** [comment]
While I do keep an eye on patches that may affect Foundation or general ObjC interop, I think runtime people like @mikeash are more suited to review.

**Comment by rposts** [comment]
@mikeash - let me know your thoughts on limiting this PR to s390x arch and whether proposed changes are acceptable. Thanks again for your insight.

**Comment by mikeash** [comment]
Yes, they look good. The cleanup I mentioned would be nice but there is no reason to make this change wait for it.

Just curious, when the comment says, "On s390x it is theoretically possible to have high bit set but in practice it is unlikely." How unlikely is it?

**Comment by rposts** [comment]
@mikeash - until recently this was not even possible on s390x with kernels supporting 4-level page table. On recent kernels with 5-level page tables, it is possible but we haven't seen any code using it.  It remains unlikely but there can be always be cases where explicit address requests can be made in this range.

**Comment by mikeash** [comment]
Interesting. Thanks for the explanation.

**Comment by gottesmm** [comment]
@rposts question. Once we get this working would it be possible to get a community bot for s390x so we stop breaking you ; )? I.e. here: https://ci-external.swift.org

**Comment by rposts** [comment]
@gottesmm there were some attempts early on but they were stalled due to T&Cs.  I am not sure what the latest status is but this is definitely something we would like to pursue.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by rposts** [comment]
@compnerd OS X platform failure above appears to be timeout related. Please let me know if something is regressed as part of this PR.  I could not find anything off the top.  thx.

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
Let's see if it works better this time.

**Comment by milseman** [comment]
@swift-ci please test os x platform

**Comment by milseman** [comment]
@swift-ci please test os x platform

**Comment by rposts** [comment]
Thanks @milseman - I assume all set to go.

File: stdlib/public/runtime/HeapObject.cpp [hunk 1]
```diff
@@ -61,9 +61,11 @@ using namespace swift;
 /// Returns true if the pointer passed to a native retain or release is valid.
 /// If false, the operation should immediately return.
 static inline bool isValidPointerForNativeRetain(const void *p) {
-#if defined(__x86_64__) || defined(__arm64__) || defined(__aarch64__) || defined(_M_ARM64)
-  // On these platforms, the upper half of address space is reserved for the
+#if defined(__x86_64__) || defined(__arm64__) || defined(__aarch64__) || defined(_M_ARM64) || defined(__s390x__)
+  // On these platforms, except s390x, the upper half of address space is reserved for the
   // kernel, so we can assume that pointer values in this range are invalid.
+  // On s390x it is theoretically possible to have high bit set but in practice
```
Comment (DougGregor) [inline stdlib/public/runtime/HeapObject.cpp]: This last commit is a bit nerve-wracking. Is there somewhere we could add an assertion for s390x so if this ever goes awry, it's easy to debug?

Comment (DougGregor) [reply stdlib/public/runtime/HeapObject.cpp]: @jckarter and @mikeash convinced me that this should be okay, so I'll worry less.

Comment (rposts) [reply stdlib/public/runtime/HeapObject.cpp]: Thanks all for helping us narrow the s390x gap!

Comment (jckarter) [reply stdlib/public/runtime/HeapObject.cpp]: Should be fine as long as swift_*alloc* never use the high half of address space. The issue here is whether a pointer to a refcountable Swift object is ever in the high half—if that's never the case, then it's OK.

Comment (milseman) [reply stdlib/public/runtime/HeapObject.cpp]: Is there a way to make this comment more descriptive then? Maybe mention that while the address space theoretically includes it, swift_alloc can never allocated objects at those addresses.

Comment (rposts) [reply stdlib/public/runtime/HeapObject.cpp]: @milseman I cannot be sure if swift_alloc can never allocate objects at those addresses.  Is there somewhere I can look to confirm?

Comment (jckarter) [reply stdlib/public/runtime/HeapObject.cpp]: By default the allocator functions just call down to the platform `malloc`, so you could look at whatever `malloc` does.

Comment (rposts) [reply stdlib/public/runtime/HeapObject.cpp]: Ah ok - yes, malloc can allocate upper range if other address space is exhausted.

How about changing the comment to read this:
```
// On s390x it is theoretically possible to have high bit set but in practice the swift/system allocators do not allocate in this range.
```

Comment (mikeash) [reply stdlib/public/runtime/HeapObject.cpp]: So it'll happen but only if you allocate ~9.2 exabytes of memory?

**Pull request merged**